src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/AbstractWorkspaceDocumentDiagnosticSource.cs

```diff
@@ -67,20 +65,24 @@ public override bool IsLiveSource()
             }
 
             var result = await lazyDiagnostics.GetValueAsync(cancellationToken).ConfigureAwait(false);
-            return (ImmutableArray<DiagnosticData>)result;
+            return result[Document.Id].ToImmutableArray();
 
-            AsyncLazy<IReadOnlyList<DiagnosticData>> GetLazyDiagnostics()
+            AsyncLazy<ILookup<DocumentId, DiagnosticData>> GetLazyDiagnostics()
             {
                 return s_projectToDiagnostics.GetValue(
                     Document.Project,
-                    _ => AsyncLazy.Create<IReadOnlyList<DiagnosticData>>(
-                        async cancellationToken => await diagnosticAnalyzerService.GetDiagnosticsForIdsAsync(
-                            Document.Project.Solution, Document.Project.Id, documentId: null,
-                            diagnosticIds: null, shouldIncludeAnalyzer,
-                            // Ensure we compute and return diagnostics for both the normal docs and the additional docs in this project.
-                            static (project, _) => [.. project.DocumentIds.Concat(project.AdditionalDocumentIds)],
-                            includeSuppressedDiagnostics: false,
-                            includeLocalDocumentDiagnostics: true, includeNonLocalDocumentDiagnostics: true, cancellationToken).ConfigureAwait(false)));
+                    _ => AsyncLazy.Create(
+                        async cancellationToken =>
+                        {
+                            var allDiagnostics = await diagnosticAnalyzerService.GetDiagnosticsForIdsAsync(
+                                Document.Project.Solution, Document.Project.Id, documentId: null,
+                                diagnosticIds: null, shouldIncludeAnalyzer,
+                                // Ensure we compute and return diagnostics for both the normal docs and the additional docs in this project.
+                                static (project, _) => [.. project.DocumentIds.Concat(project.AdditionalDocumentIds)],
+                                includeSuppressedDiagnostics: false,
+                                includeLocalDocumentDiagnostics: true, includeNonLocalDocumentDiagnostics: true, cancellationToken).ConfigureAwait(false);
+                            return allDiagnostics.Where(d => d.DocumentId != null).ToLookup(d => d.DocumentId!);
```

Comment:
>return allDiagnostics.Where(d => d.DocumentId != null).ToLookup(d => d.DocumentId!); [](http://example.com/codeflow?start=28&length=84)

Are you ever expecting to access any entry other than Document.Id? If not, why even create the lookup?

Reply:
Clarified offline.

Reply:
s_projectToDiagnostics is static Todd!